ml-agents/mlagents/trainers/agent_processor.py

```diff
@@ -1,75 +1,151 @@
-from typing import List
+from typing import List, Dict
+from collections import defaultdict
+import numpy as np
 
-from mlagents.trainers.buffer import AgentBuffer, BufferException
+from mlagents.trainers.trainer import Trainer
+from mlagents.trainers.trajectory import Trajectory, AgentExperience
+from mlagents.trainers.brain import BrainInfo
+from mlagents.trainers.tf_policy import TFPolicy
+from mlagents.trainers.action_info import ActionInfoOutputs
 
 
-class ProcessingBuffer(dict):
+class AgentProcessor:
     """
-    ProcessingBuffer contains a dictionary of AgentBuffer. The AgentBuffers are indexed by agent_id.
+    AgentProcessor contains a dictionary per-agent trajectory buffers. The buffers are indexed by agent_id.
+    Buffer also contains an update_buffer that corresponds to the buffer used when updating the model.
+    One AgentProcessor should be created per agent group.
     """
 
+    def __init__(self, trainer: Trainer, policy: TFPolicy, max_trajectory_length: int):
+        """
+        Create an AgentProcessor.
+        :param trainer: Trainer instance connected to this AgentProcessor. Trainer is given trajectory
+        when it is finished.
+        :param policy: Policy instance associated with this AgentProcessor.
+        :param max_trajectory_length: Maximum length of a trajectory before it is added to the trainer.
+        """
+        self.experience_buffers: Dict[str, List] = defaultdict(list)
+        self.last_brain_info: Dict[str, BrainInfo] = {}
+        self.last_take_action_outputs: Dict[str, ActionInfoOutputs] = defaultdict(
+            ActionInfoOutputs
+        )
+        self.stats: Dict[str, List] = defaultdict(list)
+        # Note: this is needed until we switch to AgentExperiences as the data input type.
+        # We still need some info from the policy (memories, previous actions)
+        # that really should be gathered by the env-manager.
+        self.policy = policy
+        self.episode_steps: Dict[str, int] = {}
+        self.max_trajectory_length = max_trajectory_length
+        self.trainer = trainer
+
     def __str__(self):
```

Comment:
I do not think we need a __Sttr__ method anymore.

Reply:
Removed